### DISCUSSION TITLE
How to configure target file with location of type Target in another repo

### DISCUSSION THREAD

**Discussion by castortech** [eclipse-tycho/tycho#4087] [Q&A]
First of I read about a similar question from Ed Merks [here](https://github.com/eclipse-tycho/tycho/discussions/973#discussioncomment-2804144) where it is mentioned that this is supported and it works by "reactor project".

Well I must be missing something here.

I have a common project in which I extracted some parts of the target file and included those as locations of type Target with url like:
`<location type="Target" uri="file:${project_loc:/com.castortech.iris.common.target}/slf4j.target"/>`. This worked without any issue and I built the common repo without a problem.

Now in a 2nd repo for an app, I am adding the same "fragments" to the target and now it fails.

In the log I have:
```
[INFO] Resolving target definition file:/data/jenkins/workspace/irisba-3.0/com.castortech.iris.ba.target/irisba-target.target for environments=[linux/gtk/x86_64, win32/win32/x86_64], include source mode=honor, referenced repository mode =ignore, execution environment=StandardEEResolutionHints [executionEnvironment=OSGi profile 'JavaSE-11' { source level: 11, target level: 11}], remote p2 repository 
```

and then this error a bit further:
```
[INFO] Resolving file:${project_loc:/com.castortech.iris.common.target}/slf4j.target
[WARNING] Cannot resolve ${project_loc:com.castortech.iris.common.target} target resolution might be incomplete
[ERROR] Cannot resolve target definition:
[ERROR]   N/A
[ERROR] 
```

What should I change in my configuration?

Thanks
Alain

**Comment by joeshannon** [comment]
Is the referenced project (`com.castortech.iris.common.target`) definitely part of the reactor build for the second repo? I.e. referenced explicitly or implicitly (e.g. pomless) as a `module` in the pom structure of the second repo's build?

Otherwise, as referenced in the linked discussion ([973](https://github.com/eclipse-tycho/tycho/discussions/973#discussioncomment-2804144)), it should match on either of three names:

> if either the folder name or the (maven) project name or artifact id matches the project name it should be found

(the latter "project name" in that quote refers to the value in `${project_loc:value}`)

**Reply by castortech** [reply]
Thanks @joeshannon 

First I am not a Maven expert so part of the issue might stem from there.

It wasn't directly referenced as a module but the root pom did reference the repository where this project is located. The project name is the same as the folder in the repo. 

What I ended up doing (again based on part of the referenced message) is to reference an https location hosting the file. But that location required basic authentication and that was not working. I finally hosted the file in a location without authentication and this is working.

So clearly not the optimal solution but at least it is working. I guess if I could at least re-add authentication that would make me feel better, or better yet go with the original approach that avoids even copying files as well.

**Reply by laeubi** [reply]
If you want to share a target file between (independent) maven projects one would usually deploy that file to a maven repository. `${project_loc}` only works if the target project is part of your reactor build.

**Reply by castortech** [reply]
The project being built includes the repository where the target files are located (it is in iriscommon) as shown here:

```
[DEBUG] Enter loading repository https://xxx.cloudfront.net/iriscommon-3.0
[DEBUG] Exit loading repository https://xxx.cloudfront.net/iriscommon-3.0
[DEBUG] Enter loading repository https://xxx.cloudfront.net/com.castortech.iris.noswt
[DEBUG] Exit loading repository https://xxx.cloudfront.net/com.castortech.iris.noswt
```

At the moment I put them in an accessible location via https and this is working, so at least I am able to move to the next challenge.

**Reply by laeubi** [reply]
> The project being built includes the repository where the target files are located (it is in iriscommon) as shown here:

It is not clear to me what this means... The repository does not matter as project loc only resolves to *local* projects in the reactor.
